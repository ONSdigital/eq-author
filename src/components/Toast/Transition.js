import styled from "styled-components";
import { CSSTransition } from "react-transition-group";
import PropTypes from "prop-types";

const ToastTransition = styled(CSSTransition).attrs({
  classNames: "toast"
})`
  transition: transform ${props => props.timeout}ms ease-out,
    opacity ${props => props.timeout}ms ease-out;
  transform: translateY(0);

  &.toast-enter {
    opacity: 0;
    transform: translateY(50%);
  }

  &.toast-enter.toast-enter-active {
    opacity: 1;
    transform: translateY(0);
  }

  &.toast-exit {
    opacity: 1;
    transform: translateY(0);
  }

  &.toast-exit.toast-exit-active {
    opacity: 0;
    transform: translateY(-50%);
  }
`;

ToastTransition.defaultProps = {
  in: false,
  timeout: 250
};

ToastTransition.propTypes = {
  timeout: PropTypes.number,
  children: PropTypes.element
};

export default ToastTransition;
