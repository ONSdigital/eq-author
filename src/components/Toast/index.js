import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors } from "constants/theme";
import { transparentize } from "polished";

const bgColor = transparentize(0.1, colors.darkGrey);

const StyledToast = styled.div`
  background-color: ${bgColor};
  padding: 1em;
  color: ${colors.white};
  font-size: 0.875rem;
  width: auto;
`;

export default class Toast extends React.Component {
  static propTypes = {
    timeout: PropTypes.number,
    onClose: PropTypes.func,
    children: PropTypes.node.isRequired,
    id: PropTypes.any.isRequired // eslint-disable-line react/forbid-prop-types
  };

  static defaultProps = {
    timeout: 0
  };

  constructor(props) {
    super(props);

    if (props.timeout) {
      this.timer = setTimeout(this.handleClose, props.timeout);
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  handleClose = () => {
    this.props.onClose(this.props.id);
  };

  render() {
    const { children, onClose, ...otherProps } = this.props;
    const child = React.Children.only(children);

    return (
      <StyledToast {...otherProps}>
        {React.cloneElement(child, { onClose: this.handleClose })}
      </StyledToast>
    );
  }
}
