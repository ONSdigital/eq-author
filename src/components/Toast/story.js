import React from "react";
import { storiesOf } from "@storybook/react";
import Toast from "./index";
import ToastList from "./ToastList";
import { action } from "@storybook/addon-actions";
import styled from "styled-components";
import { colors } from "constants/theme";
import Transition from "./Transition";
import { without } from "lodash";

const UndoButton = styled.button`
  background: none;
  border: none;
  color: ${colors.lightBlue};
  margin-left: 2em;
  font-size: inherit;
`;

const DeletionInfo = props => (
  <div>
    Page deleted <UndoButton onClick={action("Deleted!")}>Undo</UndoButton>
  </div>
);

const StoryContainer = styled.div`
  position: absolute;
  bottom: 0;
  text-align: center;
  width: 100%;
  margin-bottom: ${props => (props.hasMargin ? "0.5em" : "0")};
`;

const StoryInner = styled.div`
  display: inline-block;
  text-align: initial;
`;

class Story extends React.Component {
  state = {
    toasts: [Date.now()]
  };

  handleAdd = () => {
    const toasts = this.state.toasts.concat(Date.now());
    this.setState({ toasts });
  };

  handleRemove = toast => {
    const toasts = without(this.state.toasts, toast);
    this.setState({ toasts });
  };

  renderToast = (toast, i) => {
    const date = new Date(toast);

    return (
      <Toast key={toast} id={toast} onClose={this.handleRemove} timeout={5000}>
        <div>
          Toast added at: <strong>{date.toLocaleTimeString()}</strong>
        </div>
      </Toast>
    );
  };

  render() {
    return (
      <div>
        <div>
          <button type="button" onClick={this.handleAdd}>
            Add toast
          </button>
        </div>
        <StoryContainer>
          <StoryInner>
            <ToastList transition={Transition}>
              {this.state.toasts.map(this.renderToast)}
            </ToastList>
          </StoryInner>
        </StoryContainer>
      </div>
    );
  }
}

storiesOf("Toast", module)
  .add("Default", () => (
    <StoryContainer hasMargin>
      <StoryInner>
        <Toast>
          <span>Hello world!</span>
        </Toast>
      </StoryInner>
    </StoryContainer>
  ))
  .add("Auto timeout", () => (
    <StoryContainer hasMargin>
      <StoryInner>
        <Toast id="bar" timeout={2000} onClose={action("Auto-closed")}>
          <span>Hello world!</span>
        </Toast>
      </StoryInner>
    </StoryContainer>
  ))
  .add("Arbitrary content", () => (
    <StoryContainer hasMargin>
      <StoryInner>
        <Toast id="foo" timeout={2000} onClose={action("Auto-closed")}>
          <DeletionInfo />
        </Toast>
      </StoryInner>
    </StoryContainer>
  ))
  .add("List", () => (
    <StoryContainer>
      <StoryInner>
        <ToastList>
          <Toast id="1">
            <div>Hello</div>
          </Toast>
          <Toast id="2">
            <div>World</div>
          </Toast>
        </ToastList>
      </StoryInner>
    </StoryContainer>
  ))
  .add("Interactive", () => <Story />);
