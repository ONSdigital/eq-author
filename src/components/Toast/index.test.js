import React from "react";
import { shallow } from "enzyme";
import Toast from "./index";

jest.useFakeTimers();

describe("Toast", () => {
  const render = (props = {}, children = <span>hello</span>) =>
    shallow(
      <Toast id="1" {...props}>
        {children}
      </Toast>
    );

  it("should render", () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  it("should enforce a single child policy", () => {
    expect(() => {
      render({}, "hello world");
    }).toThrow();
  });

  it("should propagate onClose props to child", () => {
    const handleClose = jest.fn();
    const wrapper = render({ onClose: handleClose });
    wrapper.find("span").simulate("close");

    expect(handleClose).toHaveBeenCalled();
  });
});
