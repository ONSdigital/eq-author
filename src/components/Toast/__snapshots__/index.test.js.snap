// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render 1`] = `
<Toast__StyledToast
  id="1"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  timeout={0}
>
  <span
    onClose={[Function]}
  >
    hello
  </span>
</Toast__StyledToast>
`;
